### ISSUE TITLE
Optionally: retrieve titles of pages and URL fragment targets?

### ISSUE THREAD

**Issue by rauschma** [lycheeverse/lychee#1453]
* I’d find it useful if I could optionally track if the titles of the content I link to changes. Sometimes that means that I have to update (e.g.) the link text.
* Output could be a TSV (sorted by `URL`): `URL\tTitle`
  * If a link is broken, the `Title` would be (e.g.) an error message in square brackets.
* Users would store old output and use `diff` to compare it with new output.
* You’d have to collect, deduplicate URLs and group URLs first – to avoid repeated downloads of the same page.
* Extracting titles:
  * You’d have to extract the content of `<h1>`-`<h6>`, `<div>`, `<span>` and maybe more.
  * GitHub Markdown files (such as readmes) are problematic because they change IDs dynamically – e.g. a static `#user-content-myid` becomes `#myid` via JavaScript.

**Comment by mre** [comment]
It's an interesting idea. As with a lot of potential features of lychee, there are quite a few edge cases to consider. You mentioned some of them, but there are more:

- What if the page linked to is a text file?
- What if it's an image?
- If the page requires JavaScript to be rendered, what then? (We don't run JavaScript.)
- What's the title of a JSON API response?
- What if the user sets the request method to `PUT`?

These are just a few complications off the top of my head. There might be reasonable solutions to all of them, but the depth of the problem space is daunting. I find that this is the norm when it comes to link checking, in part because the web is a mix of different formats and standards and hard to parse automatically.

Therefore, I doubt we'll integrate that functionality. Mainly because I'd like to keep the scope of the project manageable for the team and to keep the number of false-positives at a minimum. Hope you'll understand.

**Comment by rauschma** [comment]
Sure!

* I’d only use this functionality for links to external HTML pages. I’m not sure if it would be useful elsewhere.
* If a resource does not have a useful title, the title column can simply be empty or `[OK]`. Diffing would then still work as intended.

(Feel free to close this issue if you don’t see Lychee implementing this functionality.)

**Comment by mre** [comment]
Thank you for your thoughtful suggestion and for understanding the complexity involved @rauschma.
I've decided not to implement this feature to keep lychee focused on its core link-checking functionality.

The main issue that I currently see is that there are just too many edge-cases, such as JavaScript rendering etc. for this feature to be "stable" and not lead to too many false positives down the road. Leaving the title empty in case we cannot parse it sounds like a reasonable workaround, but it would still require significant development effort for a feature that might not work reliably in many scenarios.

I believe it's a bit of an edge-case as well, as most users are primarily concerned with whether links are valid rather than tracking title changes.

I'm closing this issue now, but I appreciate the suggestion.
